satisfy/src/ConCat/BuildDictionary.hs

```diff
@@ -180,54 +187,49 @@ buildDictionary' env dflags guts evIds evar =
-     -- TODO: Investigate!
-     False &&
-     isEvVar v -- && not (isDeadBinder v)
-     -- Keep evidence that relates to free type variables in the goal.
-     -- && not (isEmptyVarSet (goalTyVars `intersectVarSet` tyCoVarsOfType (varType v))) -- see issue #20
-   -- freeIds = filter isId (uniqSetToList (exprFreeVars dict))
-   -- freeIdTys = varType <$> freeIds
-   -- goalTyVars = tyCoVarsOfType goalTy
-   reassemble (i,bnds) =
-     -- pprTrace' "buildDictionary" (ppr goalTy $$ text "-->" $$ ppr dict) $
-     -- pprTrace' "buildDictionary inScope" (ppr (fst inScope)) $
-     -- pprTrace' "buildDictionary freeIds" (ppr freeIds) $
-     -- pprTrace' "buildDictionary (bnds,freeIds)" (ppr (bnds,freeIds)) $
-     -- pprTrace' "buildDictionary dict" (ppr dict) $
-     -- either (\ e -> pprTrace' "buildDictionary fail" (ppr goalTy $$ text "-->" $$ e) res) (const res) $
+   goalTyVars = tyCoVarsOfType goalTy
+   keepVar v = not (isEmptyVarSet (goalTyVars `intersectVarSet` tyCoVarsOfType (varType v)))
+   reassemble Nothing =
+     Left (text "unsolved constraints")
+   reassemble (Just (i,bnds)) =
+     pprTrace' "buildDictionary" (ppr goalTy $$ text "-->" $$ ppr expr) $
+     pprTrace' "buildDictionary evIds" (ppr evIds) $
+     pprTrace' "buildDictionary expr" (ppr expr) $
+     either (\ e -> pprTrace' "buildDictionary fail" (ppr goalTy $$ text "-->" $$ e) res) (const res) $
      res
     where
       res | null bnds          = Left (text "no bindings")
           | notNull holeyBinds = Left (text "coercion holes: " <+> ppr holeyBinds)
-          | notNull freeIdTys  = Left (text "free id types:" <+> ppr freeIdTys)
           | otherwise          = return $ simplifyE dflags False
-                                          dict
+                                          expr
       dict =
         case bnds of
           -- Common case with single non-recursive let
           [NonRec v e] | i == v -> e
           _                     -> mkCoreLets bnds (varToCoreExpr i)
-      -- Sometimes buildDictionary' constructs bogus dictionaries with free
-      -- identifiers. Hence check that freeIds is empty. Allow for free *type*
-      -- variables, however, since there may be some in the given type as
-      -- parameters. Alternatively, check that there are no free variables (val or
-      -- type) in the resulting dictionary that were not in the original type.
-      freeIds = filterVarSet isId (exprFreeVars dict) `minusVarSet` scopedDicts
-      freeIdTys = varType <$> uniqSetToList freeIds
+      -- could optimize if these things are already variables
+      expr = if null evTypes
+             then dict
+             else mkWildCase ev evType goalTy [(DataAlt (tupleDataCon Boxed (length evIds)), evIds, dict)]
       holeyBinds = filter hasCoercionHole bnds
-      -- err doc = Left (doc $$ ppr goalTy $$ text "-->" $$ ppr dict)
 
 hasCoercionHole :: Data t => t -> Bool
 hasCoercionHole = getAny . everything mappend (mkQ mempty (Any . isHole))
```

Comment:
Oof -- this will be slow. This function exists in newer GHCs, but it doesn't in your version. (I have not tested this, so maybe I'm wrong about the "slow".) I'm not sure this check is needed, anyway: the residual constraints won't be empty if there are any holes left, I think.

Reply:
@conal Do you remember why this is there?

Reply:
I've forgotten the specific motivation. I wouldn't have thought to check for coercion holes if I didn't see them breaking something. Maybe some later GHC tweak addressed the issue.

Reply:
I've removed it for now. Should be pretty easy to see if it crops up again.